### PR TITLE
3D Beta: add support for SceneEntity arrows

### LINKED ISSUE #4315: 3D beta rendering support for SceneEntity `arrows`

### PR DESCRIPTION
**Pull request by jtbandes**
**User-Facing Changes**
None (SceneEntity support not announced yet)

**Description**
Adds support for `arrows`. Closes #4315.

https://user-images.githubusercontent.com/14237/187995637-aca877b7-84dc-4c8e-9ee8-37e7bb7cb7b9.mov

### CODE REVIEW

File: packages/studio-base/src/panels/ThreeDeeRender/renderables/primitives/RenderableArrows.ts [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+      );
+      RenderableArrows.shaftEdgesGeometry.computeBoundingSphere();
+    }
+    return RenderableArrows.shaftEdgesGeometry;
+  }
+
+  private static HeadGeometry() {
+    if (!RenderableArrows.headGeometry) {
+      RenderableArrows.headGeometry = new THREE.ConeGeometry(0.5, 1, 16);
+      RenderableArrows.headGeometry.rotateZ(-Math.PI / 2).translate(0.5, 0, 0);
```
Comment (snosenzo) [inline packages/studio-base/src/panels/ThreeDeeRender/renderables/primitives/RenderableArrows.ts:284]: also here?

File: packages/studio-base/src/panels/ThreeDeeRender/renderables/primitives/RenderableArrows.ts [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+  }
+
+  public updateSettings(settings: LayerSettingsEntity): void {
+    this.update(this.userData.entity, settings, this.userData.receiveTime);
+  }
+
+  private static ShaftGeometry() {
+    if (!RenderableArrows.shaftGeometry) {
+      RenderableArrows.shaftGeometry = new THREE.CylinderGeometry(0.5, 0.5, 1, 16);
+      RenderableArrows.shaftGeometry.rotateZ(Math.PI / 2).translate(0.5, 0, 0);
```
Comment (snosenzo) [inline packages/studio-base/src/panels/ThreeDeeRender/renderables/primitives/RenderableArrows.ts:264]: some comments on what/why these defaults are?
Also should there be consts that represent these to be shared between the shaft and head?

File: packages/studio-base/src/panels/ThreeDeeRender/renderables/primitives/RenderableArrows.ts [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+      renderer.instancedOutlineMaterial,
+    );
+    this.headOutline.frustumCulled = false;
+    this.headOutline.userData.picking = false;
+    this.add(this.headOutline);
+  }
+
+  private _ensureCapacity(numArrows: number) {
+    if (numArrows > this.maxInstances) {
+      const newCapacity = Math.ceil(numArrows * 1.5) + 16;
```
Comment (snosenzo) [inline packages/studio-base/src/panels/ThreeDeeRender/renderables/primitives/RenderableArrows.ts:113]: just curious where these numbers came from. is this just a standard for increasing the capacity for fixed size allocations?

**Review by snosenzo — APPROVED** [review]
quick question and couple comments might be helpful

File: packages/studio-base/src/panels/ThreeDeeRender/renderables/primitives/RenderableArrows.ts (hunk 3 above)
Comment (jtbandes) [reply packages/studio-base/src/panels/ThreeDeeRender/renderables/primitives/RenderableArrows.ts:113]: I got these from @jhurliman's Marker implementation. Though I do recall reading that 1.5x is better than 2x for allowing some space to be reused – but not sure that's very relevant in JS land.

Comment (jhurliman) [reply packages/studio-base/src/panels/ThreeDeeRender/renderables/primitives/RenderableArrows.ts:113]: This is the algorithm V8 uses for resizing `[]`.

**Pull request merged**